cpp/tests/shared_fcntl_mutex_test.cc

```diff
@@ -0,0 +1,516 @@
+
+// Test that acquiring read locks increases the number of fcntl readers
+TEST(SharedFcntlMutexTest, FcntlReaders) {
+  std::array<PipeComms, NCHILDREN> comms;
+  std::array<pid_t, NCHILDREN> child_pid;
+
+  ASSERT_OK(SpawnChildren(comms, child_pid));
+  constexpr std::size_t NREADLOCKS = 3;
+
+  // Base case for no locks
+  ASSERT_OK(comms[0].send(kQueryFcntlReaders, PARENT_CONTEXT));
+  ASSERT_OK(comms[0].expect("ok 0", PARENT_CONTEXT));
+
+  for (std::size_t i = 0; i < NREADLOCKS; ++i) {
+    auto thread = std::to_string(i);
+
+    // Acquire a read lock in the first child
+    ASSERT_OK(comms[0].send(kReadLock + " thread "s + thread, PARENT_CONTEXT));
+    ASSERT_OK(comms[0].expect("ok", PARENT_CONTEXT));
+
+    // Confirm number of readers
+    ASSERT_OK(comms[0].send(kQueryFcntlReaders + " thread "s + std::to_string(i),
+                            PARENT_CONTEXT));
+    ASSERT_OK(comms[0].expect("ok " + std::to_string(i + 1), PARENT_CONTEXT));
+  }
+
+  for (std::size_t i = 0; i < NREADLOCKS; ++i) {
+    auto thread = std::to_string(i);
+    // Release a read lock in the first child
+    ASSERT_OK(comms[0].send(kReadUnlock + " thread "s + thread, PARENT_CONTEXT));
+    ASSERT_OK(comms[0].expect("ok", PARENT_CONTEXT));
+
+    // Confirm number of readers
+    ASSERT_OK(comms[0].send(kQueryFcntlReaders + " thread "s + thread, PARENT_CONTEXT));
+    ASSERT_OK(
+        comms[0].expect("ok " + std::to_string(NREADLOCKS - i - 1), PARENT_CONTEXT));
+  }
+
+  ASSERT_OK(ShutdownChildren(comms, child_pid));
+}
+
+// Test that IPC locking and lock observation behaves as expected
+TEST(SharedFcntlMutexTest, InterProcessObservability) {
+  std::array<PipeComms, NCHILDREN> comms;
+  std::array<pid_t, NCHILDREN> child_pid;
+  ASSERT_OK(SpawnChildren(comms, child_pid));
+
+  // Acquire a write lock in the first process
+  ASSERT_OK(comms[0].send(kWriteLock + " thread 1"s, PARENT_CONTEXT));
+  ASSERT_OK(comms[0].expect("ok", PARENT_CONTEXT));
+
+  // Failed to acquire a write lock in the same process and thread
+  ASSERT_OK(comms[0].send(kTryWriteLock + " thread 1"s, PARENT_CONTEXT));
+  ASSERT_OK(comms[0].expect("fail", PARENT_CONTEXT));
+
+  // Failed to acquire a write lock in the same process and a different thread
+  ASSERT_OK(comms[0].send(kTryWriteLock + " thread 2"s, PARENT_CONTEXT));
+  ASSERT_OK(comms[0].expect("fail", PARENT_CONTEXT));
+
+  // Failed to acquire a write lock in a different process
+  ASSERT_OK(comms[1].send(kTryWriteLock + " thread 2"s, PARENT_CONTEXT));
+  ASSERT_OK(comms[1].expect("fail", PARENT_CONTEXT));
+
+  // Second process observes the first's write lock
+  ASSERT_OK(comms[1].send(kRequestLock + " read"s, PARENT_CONTEXT));
+  auto child_0_pid = std::to_string(child_pid[0]);
+  ASSERT_OK(comms[1].expect("fail write " + child_0_pid, PARENT_CONTEXT));
+
+  // Release write lock in the first process
+  ASSERT_OK(comms[0].send(kWriteUnlock + " thread 2"s, PARENT_CONTEXT));
```

Comment:
That's not the same thread that obtained the lock. Did you mean
```suggestion
  ASSERT_OK(comms[0].send(kWriteUnlock + " thread 1"s, PARENT_CONTEXT));
```

Reply:
yes, this has been corrected